app/proton/proton_callback_handler.py

```diff
@@ -0,0 +1,230 @@
+def process_login_case(proton_user: ProtonUser) -> ProtonCallbackResult:
+    # Try to find a SimpleLogin user registered with that proton user id
+    sl_user_with_external_id = User.get_by(
+        partner_id=PROTON_PARTNER_ID, partner_user_id=proton_user.id
+    )
+    if sl_user_with_external_id is None:
+        # We didn't find any SimpleLogin user registered with that proton user id
+        # Try to find it using the proton's e-mail address
+        sl_user = User.get_by(email=proton_user.email)
+        return get_login_strategy(proton_user, sl_user).process()
+    else:
+        # We found the SL user registered with that proton user id
+        # We're done
+        return AlreadyLinkedUserStrategy(
+            proton_user, sl_user_with_external_id
+        ).process()
+
+
+def link_user(proton_user: ProtonUser, current_user: User) -> ProtonCallbackResult:
+    current_user.partner_user_id = proton_user.id
+    current_user.partner_id = PROTON_PARTNER_ID
+
+    ensure_partner_user_exists(proton_user, current_user)
+
+    Session.commit()
+    return ProtonCallbackResult(
+        redirect_to_login=False,
+        redirect=url_for("dashboard.setting"),
+        flash_category="success",
+        flash_message="Account successfully linked",
+        user=current_user,
+    )
+
+
+def process_link_case(
+    proton_user: ProtonUser, current_user: User
+) -> ProtonCallbackResult:
+    # Try to find a SimpleLogin user linked with this Proton account
+    sl_user_linked_to_proton_account = User.get_by(
+        partner_id=PROTON_PARTNER_ID, partner_user_id=proton_user.id
+    )
+    if sl_user_linked_to_proton_account is None:
+        # There is no SL user linked with the proton email. Proceed with linking
+        return link_user(proton_user, current_user)
+    else:
+        # There is a SL user registered with the proton email. Check if is the current one
+        if sl_user_linked_to_proton_account.id == current_user.id:
+            # It's the same user. No need to do anything
+            return ProtonCallbackResult(
+                redirect_to_login=False,
+                redirect=url_for("dashboard.setting"),
+                flash_category="success",
+                flash_message="Account successfully linked",
+                user=current_user,
+            )
+        else:
+            # It's a different user. Unlink the other account and link the current one
+            sl_user_linked_to_proton_account.partner_id = None
+            sl_user_linked_to_proton_account.partner_user_id = None
+            other_partner_user = PartnerUser.get_by(
+                user_id=sl_user_linked_to_proton_account.id,
+                partner_id=PROTON_PARTNER_ID,
+            )
+            if other_partner_user is not None:
+                PartnerUser.delete(other_partner_user.id)
+
+            return link_user(proton_user, current_user)
+
+
+def process_case(
+    proton_user: ProtonUser, action: Action, current_user: Optional[User]
```

Comment:
Nits: should `current_user` be renamed to `sl_user` as above methods use this naming?

Reply:
In this case, the `current_user` references to the "currently logged-in user".

When the code from this function branches depending on the action, it either remains as `current_user` (because it represents the currently logged-in user, or it disappears and then the retrieving process from the DB for obtaining the `sl_user` happens.